fastscanner/adapters/candle/parquet.py

```diff
@@ -0,0 +1,185 @@
+
+
+    def _dataset_path(self, symbol: str, freq: str) -> str:
+        return os.path.join(self.CACHE_DIR, f"{symbol}_{freq}.parquet")
+
+    def _save_cache(self, symbol: str, freq: str, df: pd.DataFrame):
+        if df.empty:
+            return
+
+        path = self._dataset_path(symbol, freq)
+        os.makedirs(os.path.dirname(path), exist_ok=True)
+
+        df = df.reset_index()
+
+        df["date"] = df[CandleCol.DATETIME].dt.strftime(
+            "%Y-%m-%d" if "min" in freq else "%Y-%m"
+        )
+
+        df[CandleCol.DATETIME] = df[CandleCol.DATETIME].dt.tz_convert("UTC").dt.tz_localize(None)
+
+        if os.path.exists(path):
+            existing_df = pq.read_table(path).to_pandas()
+            combined_df = pd.concat([existing_df, df]).drop_duplicates(subset=[CandleCol.DATETIME], keep="last")
+        else:
+            combined_df = df
+
+        table = pa.Table.from_pandas(combined_df, preserve_index=False)
+        pq.write_table(table, path)
+
+    def _current_range_path(self, symbol: str) -> str:
+        return os.path.join(self.CACHE_DIR, f"{symbol}_current_ranges.json")
+
+    def _load_current_range(self, symbol: str, freq: str) -> List[Tuple[date, date]]:
+        try:
+            with open(self._current_range_path(symbol)) as f:
+                ranges = json.load(f)
+                return [
+                    (date.fromisoformat(start), date.fromisoformat(end))
+                    for start, end in ranges.get(freq, [])
+                ]
+        except (FileNotFoundError, KeyError):
+            return []
+
+    def _save_current_range(self, symbol: str, freq: str, new_ranges: List[Tuple[date, date]]):
+        path = self._current_range_path(symbol)
+        os.makedirs(os.path.dirname(path), exist_ok=True)
+
+        try:
+            with open(path) as f:
+                ranges = json.load(f)
+        except FileNotFoundError:
+            ranges = {}
+
+        existing = self._load_current_range(symbol, freq)
+        all_ranges = existing + new_ranges
+        all_ranges = sorted(all_ranges)
+
+        merged = []
+        for rng in all_ranges:
+            if not merged:
+                merged.append(rng)
+            else:
+                last_start, last_end = merged[-1]
+                curr_start, curr_end = rng
+                if curr_start <= last_end + timedelta(days=1):
+                    merged[-1] = (last_start, max(last_end, curr_end))
+                else:
+                    merged.append(rng)
+
+        ranges[freq] = [(s.isoformat(), e.isoformat()) for s, e in merged]
+        with open(path, "w") as f:
+            json.dump(ranges, f)
+
+    def _missing_ranges(self, symbol: str, start: date, end: date, freq: str) -> List[Tuple[date, date]]:
+        existing_ranges = self._load_current_range(symbol, freq)
+        if not existing_ranges:
+            return [(start, end)]
+
+        missing = []
+        current = start
+        for s, e in sorted(existing_ranges):
```

Comment:
```suggestion
        for s, e in existing_ranges:
```

They're already getting sorted before write.